### PR TITLE
Adds xct_specific_matcher opt-in rule

### PR DESCRIPTION
**Pull request by ornithocoder**
Implements #1874.

### CODE REVIEW

File: Tests/SwiftLintFrameworkTests/RulesTests.swift [hunk 1]
```diff
@@ -461,6 +461,10 @@ class RulesTests: XCTestCase {
         verifyRule(XCTFailMessageRule.description)
     }
 
+    func testXCTSpecificMatcher() {
```
Comment (marcelofabri) [inline Tests/SwiftLintFrameworkTests/RulesTests.swift]: this is already tested on `XCTSpecificMatcherRuleTests`

**Comment by marcelofabri** [comment]
Some of the violations in oss-check are false positives, such as `XCTAssertEqual(manager?.isReachableOnWWAN, false)`, because `XCTAssetFalse` takes a `Bool`, not `Bool?`.

@ornithocoder Do you think it would make sense to try to catch some obvious cases where an optional is used (e.g. optional chaining) and not generate violations? IMO it would make the rule much more usable in the real world.

**Comment by ornithocoder** [comment]
@marcelofabri I can look for "?" on the string. Should I look for "!" as well to make sure it's not being force unwrapped?

The only problem is that the rule will "understand" certain optionals but not all of them. The way it is now, it always throws a violation, no matter if it has a implicit or explicit optional.

For `XCTAssertEqual(manager?.isReachableOnWWAN, false)` the developer can comply to the `xct_specific_matcher` rule by using:

* `XCTAssertFalse(manager!.isReachableOnWWAN)`, or
* `XCTAssertFalse(manager?.isReachableOnWWAN == .some(true))`.

I'll try something with the "?" to see how it goes :-)

**Comment by marcelofabri** [comment]
I think checking for `?` should be enough. 

I agree that it won't solve the issue in all cases, but I think it solves a common pattern when testing methods that return optional or `throws`, for example: https://github.com/kickstarter/ios-oss/blob/ccde34f1b03ddd5ab327254e78ae7473b7473bda/KsApi/models/UserTests.swift#L52:5.

I think it'd be important to handle it if possible because the other alternatives are way worse IMO. Force unwrapping means that the test would crash instead of failing, which can be hard to diagnose. And comparing to `.some(true)` is just ugly 😅

**Comment by marcelofabri** [comment]
An important detail though is that checking for `?` shouldn't be done for `XCTAsset(Not)Nil`.

**Comment by ornithocoder** [comment]
@marcelofabri I noticed something really *interesting* when implementing the latest patch. When the argument is optional, it doesn't show on the structure.

On the example below, it's not optional:

```swift
XCTAssertEqual(manager.isReachableOnWWAN, false)
```

so both arguments show on the structure

```json
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.substructure" : [
    {
      "key.namelength" : 14,
      "key.nameoffset" : 0,
      "key.length" : 48,
      "key.substructure" : [
        {
          "key.namelength" : 0,
          "key.nameoffset" : 0,
          "key.length" : 25,
          "key.bodyoffset" : 15,
          "key.offset" : 15,
          "key.bodylength" : 25,
          "key.kind" : "source.lang.swift.expr.argument"
        },
        {
          "key.namelength" : 0,
          "key.nameoffset" : 0,
          "key.length" : 5,
          "key.bodyoffset" : 42,
          "key.offset" : 42,
          "key.bodylength" : 5,
          "key.kind" : "source.lang.swift.expr.argument"
        }
      ],
      "key.name" : "XCTAssertEqual",
      "key.bodyoffset" : 15,
      "key.bodylength" : 32,
      "key.kind" : "source.lang.swift.expr.call",
      "key.offset" : 0
    }
  ],
  "key.offset" : 0,
  "key.length" : 49
}
```

But on this example, it is:

```swit
XCTAssertEqual(manager?.isReachableOnWWAN, false)
```

and the optional argument is completely gone:

```json
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.substructure" : [
    {
      "key.namelength" : 14,
      "key.nameoffset" : 0,
      "key.length" : 49,
      "key.substructure" : [
        {
          "key.namelength" : 0,
          "key.nameoffset" : 0,
          "key.length" : 5,
          "key.bodyoffset" : 43,
          "key.offset" : 43,
          "key.bodylength" : 5,
          "key.kind" : "source.lang.swift.expr.argument"
        }
      ],
      "key.name" : "XCTAssertEqual",
      "key.bodyoffset" : 15,
      "key.bodylength" : 33,
      "key.kind" : "source.lang.swift.expr.call",
      "key.offset" : 0
    }
  ],
  "key.offset" : 0,
  "key.length" : 50
}
```

The following two examples are force-unwrapping the optional argument. Therefore, it's shown on the on the structure.

```swift
XCTAssertEqual(manager!.isReachableOnWWAN, false)
```

```json
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.substructure" : [
    {
      "key.namelength" : 14,
      "key.nameoffset" : 0,
      "key.length" : 49,
      "key.substructure" : [
        {
          "key.namelength" : 0,
          "key.nameoffset" : 0,
          "key.length" : 26,
          "key.bodyoffset" : 15,
          "key.offset" : 15,
          "key.bodylength" : 26,
          "key.kind" : "source.lang.swift.expr.argument"
        },
        {
          "key.namelength" : 0,
          "key.nameoffset" : 0,
          "key.length" : 5,
          "key.bodyoffset" : 43,
          "key.offset" : 43,
          "key.bodylength" : 5,
          "key.kind" : "source.lang.swift.expr.argument"
        }
      ],
      "key.name" : "XCTAssertEqual",
      "key.bodyoffset" : 15,
      "key.bodylength" : 33,
      "key.kind" : "source.lang.swift.expr.call",
      "key.offset" : 0
    }
  ],
  "key.offset" : 0,
  "key.length" : 50
}
```

```swift
XCTAssertEqual((manager?.isReachableOnWWAN)!, false)
```

```json
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.substructure" : [
    {
      "key.namelength" : 14,
      "key.nameoffset" : 0,
      "key.length" : 52,
      "key.substructure" : [
        {
          "key.namelength" : 0,
          "key.nameoffset" : 0,
          "key.length" : 29,
          "key.bodyoffset" : 15,
          "key.offset" : 15,
          "key.bodylength" : 29,
          "key.kind" : "source.lang.swift.expr.argument"
        },
        {
          "key.namelength" : 0,
          "key.nameoffset" : 0,
          "key.length" : 5,
          "key.bodyoffset" : 46,
          "key.offset" : 46,
          "key.bodylength" : 5,
          "key.kind" : "source.lang.swift.expr.argument"
        }
      ],
      "key.name" : "XCTAssertEqual",
      "key.bodyoffset" : 15,
      "key.bodylength" : 36,
      "key.kind" : "source.lang.swift.expr.call",
      "key.offset" : 0
    }
  ],
  "key.offset" : 0,
  "key.length" : 53
}
```

I've used this on the implementation not to trigger the violation for equal true/false when the argument is optional.

**Comment by ornithocoder** [comment]
@marcelofabri maybe i'm too tired today, but i'm having troubles to get the argument as string.

For instance, for:

```
foo(a.bar(b)?.c, false)
```

I want `a.bar(b)?.c`, but all I get from offset and length is `a.bar(b)`

```json
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.substructure" : [
    {
      "key.namelength" : 3,
      "key.nameoffset" : 0,
      "key.length" : 23,
      "key.substructure" : [
        {
          "key.namelength" : 5,
          "key.nameoffset" : 4,
          "key.length" : 8,
          "key.name" : "a.bar",
          "key.bodyoffset" : 10,
          "key.bodylength" : 1,
          "key.kind" : "source.lang.swift.expr.call",
          "key.offset" : 4
        },
        {
          "key.namelength" : 0,
          "key.nameoffset" : 0,
          "key.length" : 5,
          "key.bodyoffset" : 17,
          "key.offset" : 17,
          "key.bodylength" : 5,
          "key.kind" : "source.lang.swift.expr.argument"
        }
      ],
      "key.name" : "foo",
      "key.bodyoffset" : 4,
      "key.bodylength" : 18,
      "key.kind" : "source.lang.swift.expr.call",
      "key.offset" : 0
    }
  ],
  "key.offset" : 0,
  "key.length" : 24
}
```

**Comment by marcelofabri** [comment]
@ornithocoder I wouldn't rely on that behavior (optional chaining not showing up in structure). Check https://github.com/realm/SwiftLint/issues/1785. It was a bug in SourceKit ([SR-5717](https://bugs.swift.org/browse/SR-5717)) which was fixed by apple/swift#11542,

**Comment by ornithocoder** [comment]
Hmmm. Any suggestion of what should I do in this case?

1) Use regular expressions (which I tried before, according the commit history. I don't remember why I walked away from it)
2) Don't handle optionals for now
3) Leave this PR open until we can use apple/swift#11542

Thanks :-)

**Comment by marcelofabri** [comment]
Would it make sense to change the implementation based on Swift version?

For example:

## Swift >= 4.1 

Optional chaining is returned in the structure, so we could use that to get the text and check for `?`.

## Swift < 4.1

Optional chaining is NOT returned in the structure, so we could only validate calls with (at least) 2 arguments in it.

**Comment by ornithocoder** [comment]
it sounds good. I'll do that.

**Comment by ornithocoder** [comment]
I'll have to check why the performance looks so bad. Maybe I should check if the class is a XCTestCase  subclass (but then it would not work for people using XCTest matchers with Quick/Specta/etc...).

**Comment by marcelofabri** [comment]
@ornithocoder I think it'd be ok to test if it's a subclass of `XCTestCase` or `QuickSpec`.

I'd also not trust *too* much on oss-check running on Circle CI. You can run `script/oss-check` locally to get more accurate results.

**Comment by ornithocoder** [comment]
Hi @marcelofabri, sorry about that. I'll rebase and update the title.

**Comment by ornithocoder** [comment]
@marcelofabri this is "ready". I was trying to use it and found some problems that I don't think it's possible to handle. For instance

```swift
var foo: Foo?

foo = // ...

XCTAssertEqual(foo.isValid, true)
```

`foo` is optional and there's no way to know it's optional, therefore the rule suggests to use `XCTAssertTrue` (which is incorrect, since `XCTAssertTrue` doesn't take optional arguments).

**Comment by marcelofabri** [comment]
Merged in #2528 🚀

**Pull request closed without merging**